src/api/habits/adapters/json.ts

```diff
@@ -26,7 +30,75 @@ const jsonHabitServiceFactory = (): HabitService => {
         }))
       );
 
-  return { addHabit, getHabits };
+  const addEntry = (
+    goals: GoalWithHabitHistory[],
+    habitId: string,
+    date: Date = new Date(),
+    quantity: number = 1
+  ) => {
+    try {
+      const updatedGoals = goals.map((goal) => ({
+        ...goal,
+        habits: goal.habits.reduce((habitsAcc, currentHabit) => {
+          let habit = currentHabit;
+
+          if (habit.id === habitId) {
+            // Add an entry here
+            const newEntries = habit.entries.concat([
+              { completionDate: date, quantity },
+            ]);
+            habit = {
+              ...habit,
+              entries: newEntries,
+            };
+          }
+
+          // Otherwise just return this habit as-is
+          return habitsAcc.concat(habit);
+        }, [] as HabitWithHistory[]),
+      }));
+
+      return Promise.resolve(updatedGoals);
+    } catch (error) {
+      return Promise.reject(error);
+    }
+  };
+
+  const removeEntry = (
+    goals: GoalWithHabitHistory[],
+    habitId: string,
+    date: Date = new Date()
```

Comment:
It appears you may be using date as the primary key/identifier. It'd be preferable to use something else - more stable too. Perhaps you should generate an id on entry addition instead.